pkg/reconciler/taskrun/taskrun_expired.go

```diff
@@ -0,0 +1,181 @@
+package taskrun
+
+import (
+	"fmt"
+	"time"
+
+	apispipeline "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/controller"
+	"knative.dev/pkg/apis"
+)
+
+func (tc *Reconciler) AddTaskRun(obj interface{}) {
+	tr := obj.(*apispipeline.TaskRun)
+	klog.V(4).Infof("Adding TaskRun %s/%s", tr.Namespace, tr.Name)
+
+	if tr.DeletionTimestamp == nil && taskRunCleanup(tr) {
+		tc.TrEnqueue(tr)
+	}
+}
+
+func (tc *Reconciler) UpdateTaskRun(old, cur interface{}) {
+	tr := cur.(*apispipeline.TaskRun)
+	klog.V(4).Infof("Updating TaskRun %s/%s", tr.Namespace, tr.Name)
+
+	if tr.DeletionTimestamp == nil && taskRunCleanup(tr) {
+		tc.TrEnqueue(tr)
+	}
+}
+
+func (tc *Reconciler) TrEnqueue(tr *apispipeline.TaskRun) {
+	klog.V(4).Infof("Add TaskRun %s/%s to cleanup", tr.Namespace, tr.Name)
+	key, err := controller.KeyFunc(tr)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", tr, err))
+		return
+	}
+
+	tc.queue.Add(key)
+}
+
+func (tc *Reconciler) TrEnqueueAfter(tr *apispipeline.TaskRun, after time.Duration) {
+	key, err := controller.KeyFunc(tr)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", tr, err))
+		return
+	}
+
+	tc.queue.AddAfter(key, after)
+}
+
+// processTaskRun will check the TaskRun's state and TTL and delete the TaskRun when it
+// finishes and its TTL after finished has expired. If the TaskRun hasn't finished or
+// its TTL hasn't expired, it will be added to the queue after the TTL is expected
+// to expire.
+// This function is not meant to be invoked concurrently with the same key.
+func (tc *Reconciler) processTaskRunExpired(namespace, name string, tr *apispipeline.TaskRun) error {
```

Comment:
I'm curious, do you think it's possible to use the same logic for TaskRuns and PipelineRuns, or do they definitely need to be distinct?

Reply:
Because processTaskRunExpired needs to call HasPipelineRunOwnerReference but  processTaskRunExpired not, and tc.processTrTTL calls taskRunCleanup which is different from pipelineRunCleanup. So I would like to implement this separately :)